docs/raw/push-notification-server.md

```diff
@@ -0,0 +1,672 @@
+  - [Sending a push notification](#sending-a-push-notification)
+  - [Flow](#flow)
+    - [Registration process](#registration-process)
+    - [Sending a notification](#sending-a-notification)
+    - [Receiving a notification](#receiving-a-notification)
+  - [Protobuf description](#protobuf-description)
+    - [PushNotificationRegister](#pushnotificationregister)
+    - [PushNotificationPreferences](#pushnotificationpreferences)
+    - [PushNotificationDeviceToken](#pushnotificationdevicetoken)
+    - [PushNotificationOptions](#pushnotificationoptions)
+    - [PushNotificationPreferencesAcknowledgment](#pushnotificationpreferencesacknowledgement)
+    - [ContactCodeAdvertisement](#contactcodeadvertisement)
+    - [PushNotificationAdvertisementInfo](#pushnotificationadvertisementinfo)
+    - [PushNotificationQuery](#pushnotificationquery)
+    - [PushNotificationQueryInfo](#pushnotificationqueryinfo)
+    - [PushNotificationQueryResponse](#pushnotificationqueryresponse)
+    - [PushNotification](#pushnotification)
+    - [PushNotificationRequest](#pushnotificationrequest)
+    - [PushNotificationAcknowledgement](#pushnotificationacknowledgement)
+  - [Partitioned topic](#partitioned-topic)
+  - [FAQ](#faq)
+  - [Changelog](#changelog)
+    - [Version 0.1](#version-01)
+  - [Copyright](#copyright)
+
+## Reason
+
+Push notifications for IOS devices and for some android device can only be
+implemented relying on [APN service](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html#//apple_ref/doc/uid/TP40008194-CH8-SW1) for IOS or [Firebase](https://firebase.google.com/) for 
+Android devices that do not support background services or that often kill the 
+background service.
+Therefore Status provides a set of Push notification services that can be used
+to achieve this functionality.
+
+Because this can't be safely implemented in a privacy preserving manner, clients
+MUST be given an option not to enable push notifications and not to send push notifications.
+
+## Requirements
+
+Status needs to run their own [gorush](https://github.com/appleboy/gorush) 
+publicly accessible server for sending the actual notification. This cannot be community run as it requires sharing private credentials that only status has access to.
+
+## Registering with the push notification service
+
+A client MAY register with a Push Notification service of their choice.
+
+[//]: # (We can have a separate partitioned topic scheme, or individual based on PK)
+
+A message MUST be sent to the [partitioned topic](#partitioned-topic) for the
+public key of the node, encrypted with this key.
+
+The content of the message MUST contain the following [protobuf record](https://developers.google.com/protocol-buffers/):
+
+```protobuf
+
+message PushNotificationOptions {
+  boolean enabled = 1;
+  repeated string allowed_user_list = 2;
+  repeated string block_user_list = 3;
```

Comment:
Is this supposed to be appending? Could this get out of hand in terms of the number of users allowed/blocked? (I suppose that can be taken care of later)

Reply:
This is straight replacing, given that this is a fairly rare message and we have a 1MB limit, I think we are safe in terms of size, if size will be a concern we can always implement a partial update.